test/cpp/qps/client_async.cc

```diff
@@ -949,10 +949,182 @@ std::unique_ptr<Client> CreateAsyncClient(const ClientConfig& config) {
       return nullptr;
   }
 }
+
 std::unique_ptr<Client> CreateGenericAsyncStreamingClient(
     const ClientConfig& args) {
   return std::unique_ptr<Client>(new GenericAsyncStreamingClient(args));
 }
 
+struct CallbackClientRpcData {
+  CallbackClientRpcData()
+      : thread_mu_(), thread_cv_(), rpc_done_(false), alarm_(nullptr) {}
+
+  ~CallbackClientRpcData() { delete alarm_; }
+
+  std::mutex thread_mu_;
+  std::condition_variable thread_cv_;
+  bool rpc_done_;
+  Alarm* alarm_;
+};
+
+class CallbackClient
+    : public ClientImpl<BenchmarkService::Stub, SimpleRequest> {
+ public:
+  CallbackClient(const ClientConfig& config)
+      : ClientImpl<BenchmarkService::Stub, SimpleRequest>(
+            config, BenchmarkStubCreator) {
+    num_threads_ = NumThreads(config);
+    ctxs_.resize(num_threads_);
+    threads_done_ = 0;
+    SetupLoadTest(config, num_threads_);
+  }
+
+  virtual ~CallbackClient() {}
+
+  virtual bool InitThreadFuncImpl(size_t thread_idx) { return true; }
+
+  virtual bool ThreadFuncImpl(Thread* t, size_t thread_idx) = 0;
+
+ protected:
+  size_t num_threads_;
+  std::mutex mu_;
+  std::condition_variable cv_;
+  std::vector<CallbackClientRpcData*> ctxs_;
+  size_t threads_done_;
+
+  void NotifyThread(size_t thread_idx) {
+    std::mutex* thread_mu = &ctxs_[thread_idx]->thread_mu_;
+    std::condition_variable* thread_cv = &ctxs_[thread_idx]->thread_cv_;
+    std::lock_guard<std::mutex> thread_lock(*thread_mu);
+    ctxs_[thread_idx]->rpc_done_ = true;
+    thread_cv->notify_one();
+  }
+
+  void IssueCallbackRpc(size_t thread_idx, Thread* t) {
+    if (!InitThreadFuncImpl(thread_idx)) {
+      return;
+    }
+
+    if (!closed_loop_) {
+      gpr_timespec next_issue_time = NextIssueTime(thread_idx);
+      // Start an alarm call back to run the internal callback after
+      // next_issue_time
+      delete ctxs_[thread_idx]->alarm_;
+      ctxs_[thread_idx]->alarm_ = new Alarm;
+      Alarm* alarm = ctxs_[thread_idx]->alarm_;
+      alarm->experimental().Set(
+          next_issue_time,
+          [this, t, thread_idx](bool ok) { ThreadFuncImpl(t, thread_idx); });
+    } else {
+      ThreadFuncImpl(t, thread_idx);
+    }
+  }
+
+ private:
+  int NumThreads(const ClientConfig& config) {
+    int num_threads = config.async_client_threads();
+    if (num_threads <= 0) {  // Use dynamic sizing
+      num_threads = cores_;
+      gpr_log(GPR_INFO, "Sizing async client to %d threads", num_threads);
```

Comment:
Nit: change "async" to "callback" client so that the log statement isn't outputting exactly the same thing as the log statement for the async client in its version of `NumThreads`